### DISCUSSION TITLE
Animation not working with <Link> tag

### DISCUSSION THREAD

**Discussion by ByAnthony** [vercel/next.js#68080] [Help] [answered]
## Summary

I am having some issues with css animation. I have a fade in animation for a menu component at the top of every pages of my app. 

The `<Link>` tag used to navigate to another page seems to break the animation and nothing happens. If i use the `<a>` tag, the animation on the menu component works.

Does the issue come from the `<Link>` tag or am i missing something?

## System information
OS: macOS
Browser: Chrome
Version of Next.js: 14.2.4
Version of Node.js: 18.18.1

**Comment by icyJoseph** [comment]
Show us some code? How's the Link being used?

**Reply by ByAnthony** [reply]
Hi @icyJoseph 👋 That is my footer link for example:
```
<div className={STYLES["map-link"]}>
  <Link href="/about-us">About Us</Link>
</div>
```
[Source](https://github.com/ByAnthony/newzealandtunnellers/blob/a9960f4d1386419bcdb6b99053eeb83adf02efc5/components/Footer/Footer.tsx#L24)

If i click that link, the animation on the next page doesn't work, but work with a `<a>` tag 🙃

**Reply by icyJoseph** [reply]
Ok, let me see if I understand.

The animation is failing on the next page... but it works with an `a` tag.

That sounds like a problem with SPA. Using an `a` tag, unloads the page, and then reloads, so the animations re-trigger.

With the Link component, then the page is not unloaded, so your animation is not triggered.

Which component is supposed to animate? Could you link to it?

**Reply by ByAnthony** [reply]
It is supposed to animate the [Menu component](https://github.com/ByAnthony/newzealandtunnellers/blob/a9960f4d1386419bcdb6b99053eeb83adf02efc5/components/Menu/Menu.tsx#L57) with a basic fade-in (the [css animation](https://github.com/ByAnthony/newzealandtunnellers/blob/a9960f4d1386419bcdb6b99053eeb83adf02efc5/app/globals.scss#L104-L115)).

**Reply by icyJoseph** [reply, marked as answer]
Alright, hear me out here, I want you to try this:

In the Menu Component:

```jsx
import { usePathname } from 'next/navigation'
```

And then in the div that receives the animation:

```jsx
    <div key={pathname} data-testid="menu" className={`${STYLES.menu} fade-in`}>
```

Try that, and report back what you see.

**Reply by ByAnthony** [reply]
this works, thanks 🎉 

Can you explain how assigning the key prop to the current url pathname makes a difference? does it understand it changes page and needs to unload/reload?

**Reply by icyJoseph** [reply]
The key is a way to tell React that an instance hasn't changed. Typically we do this in lists, so that when the list changes, React knows it can re-use DOM nodes for the unchanged elements, rather than creating them from scratch.

However this can be used to reset an instance.

- https://react.dev/reference/react/useState#resetting-state-with-a-key
- https://react.dev/learn/preserving-and-resetting-state

> You can reset a component’s state by passing a different key to a component. In this example, the Reset button changes the version state variable, which we pass as a key to the Form. When the key changes, React re-creates the Form component (and all of its children) from scratch, so its state gets reset.

So this is a syntactically cheap way to tell React to trash a DOM node, and all of its children, and recreate it in the DOM, which in turn executes CSS animations on the new DOM element.

This could be good enough for you... or maybe you want to look into other techniques...

The reason why it didn't do this before, is that, when React renders the component tree, by calling your functions, it sees that some parts of the result matches the current UI, in particular the Menu items, so it doesn't do any changes to the DOM, which is why React is pretty fast.

Please don't forget to mark as answered.
